src/MerchantCenter/MerchantStatuses.php

```diff
@@ -251,53 +251,67 @@ protected function fetch_issues( string $type = null, int $per_page = 0, int $pa
+	 * Return only the valid statuses from a provided array of Google IDs. Invalid statuses:
 	 * - Aren't synced by this extension (invalid ID format), or
 	 * - Map to products no longer in WooCommerce (deleted or uploaded by a previous connection).
-	 * Also populates the $product_name_lookup used in refresh_product_issues()
+	 * Also populates the product_data_lookup used in refresh_product_issues()
+	 *
+	 * @param string[] $google_ids
 	 *
-	 * @param Shopping_Product_Status[] $mc_statuses
 	 * @return Shopping_Product_Status[] Statuses found to be valid.
 	 */
-	protected function filter_valid_statuses( array $mc_statuses ): array {
+	protected function filter_valid_statuses( array $google_ids ): array {
+		/** @var Merchant $merchant */
+		$merchant = $this->container->get( Merchant::class );
 		/** @var ProductHelper $product_helper */
-		$product_helper = $this->container->get( ProductHelper::class );
+		$product_helper      = $this->container->get( ProductHelper::class );
+		$visibility_meta_key = $this->prefix_meta_key( ProductMetaHandler::KEY_VISIBILITY );
 
-		return array_values(
-			array_filter(
-				$mc_statuses,
-				function( $product ) use ( $product_helper ) {
-					$wc_product_id = $product_helper->get_wc_product_id( $product->getProductId() );
-					// Skip products not synced by this extension.
-					if ( ! $wc_product_id ) {
-						return false;
-					}
+		$valid_statuses = [];
+		foreach ( $merchant->get_productstatuses_batch( $google_ids )->getEntries() as $response_entry ) {
+			$mc_product_status = $response_entry->getProductStatus();
+			if ( ! $mc_product_status ) {
+				do_action(
+					'gla_debug_message',
+					sprintf( 'A Google ID was not found in Merchant Center.' ),
+					__METHOD__
+				);
+				continue;
+			}
+			$mc_product_id = $mc_product_status->getProductId();
+			$wc_product_id = $product_helper->get_wc_product_id( $mc_product_id );
+			// Skip products not synced by this extension.
+			if ( ! $wc_product_id ) {
+				continue;
+			}
 
-					// Product previously found/validated.
-					if ( ! empty( $this->product_data_lookup[ $wc_product_id ] ) ) {
-						return true;
-					}
+			// Product previously found/validated.
+			if ( ! empty( $this->product_data_lookup[ $wc_product_id ] ) ) {
+				$valid_statuses[] = $response_entry->getProductStatus();
+				continue;
+			}
 
-					$wc_product = wc_get_product( $wc_product_id );
-					// Skip products that are no longer in WooCommerce.
-					if ( ! $wc_product ) {
-						do_action(
-							'woocommerce_gla_debug_message',
-							sprintf( 'Merchant Center product %s not found in this WooCommerce store.', $product->getProductId() ),
-							__METHOD__ . ' in remove_invalid_statuses()',
-						);
-						return false;
-					}
+			$wc_product = get_post( $wc_product_id );
+			if ( ! $wc_product || 'product' !== substr( $wc_product->post_type, 0, 7 ) ) {
+				// Should never reach here since the products IDS are retrieved from postmeta.
```

Comment:
It's _possible_ that an ID retrieved from Merchant Center is actually a `post` that's not a `product`, so I preferred to skip it instead of counting a non-product.